src/tools/w3dviewer/datatreeview.cpp

```diff
@@ -0,0 +1,680 @@
+                        document->SetParticleEmitter((ParticleEmitterClass *)robj, true, true);
+                        Ref_Ptr_Release(robj);
+                    } break;
+                    default: {
+                        RenderObjClass *robj = GetRenderObj(item);
+                        document->SetRenderObject(robj, true, true, false);
+                        Ref_Ptr_Release(robj);
+                    } break;
+                }
+
+                CMainFrame *frame = ((CMainFrame *)AfxGetMainWnd());
+
+                if (frame != nullptr) {
+                    frame->UpdateMenus(info->m_type);
+
+                    if (info->m_type == ASSET_TYPE_AGGREGATE) {
+                        frame->UpdateEmitterMenu();
+                    } else {
+                        EnableMenuItem(GetSubMenu(::GetMenu(frame->m_hWnd), 3), 3, MF_BYPOSITION | MF_DISABLED | MF_GRAYED);
+
+                        while (RemoveMenu(frame->m_subMenu, 0, MF_BYPOSITION)) {
+                        }
+                    }
+                }
+            }
+        }
+
+        return;
+    }
+
+    CW3DViewDoc *document = ((CW3DViewDoc *)m_pDocument);
+
+    if (document != nullptr) {
+        document->SetRenderObject(nullptr, true, true, false);
+        CMainFrame *frame = ((CMainFrame *)AfxGetMainWnd());
+
+        if (frame != nullptr) {
+            frame->UpdateMenus(ASSET_TYPE_NONE);
+        }
+    }
+}
+
+void SetLODLevel(RenderObjClass *robj)
+{
+    if (robj != nullptr) {
+        for (int i = 0; i < robj->Get_Num_Sub_Objects(); i++) {
+            RenderObjClass *o = robj->Get_Sub_Object(i);
+
+            if (o != nullptr) {
+                SetLODLevel(o);
+                o->Release_Ref();
+            }
+        }
+
+        if (robj->Class_ID() == RenderObjClass::CLASSID_HLOD) {
+            robj->Set_LOD_Level(robj->Get_LOD_Count() - 1);
+        }
+    }
+}
+
+RenderObjClass *CDataTreeView::GetRenderObj(HTREEITEM item)
+{
+    RenderObjClass *robj = nullptr;
+    AssetInfoClass *info = (AssetInfoClass *)GetTreeCtrl().GetItemData(item);
+
+    if (info != nullptr) {
+        Ref_Ptr_Set(robj, info->m_renderObj);
+
+        if (robj == nullptr) {
+            if (info->m_type == ASSET_TYPE_TEXTURE) {
+                if (info->m_texture != nullptr) {
+                    robj = new Bitmap2D(info->m_texture, 0.5f, 0.5f, true, false, false, true);
+                }
+            }
+
+            if (robj == nullptr) {
+                robj = W3DAssetManager::Get_Instance()->Create_Render_Obj(info->m_name);
+            }
+        }
+
+        if (robj != nullptr) {
+            CW3DViewDoc *document = GetCurrentDocument();
+
+            if (!document->m_scene->Get_Auto_Switch_LOD()) {
+                SetLODLevel(robj);
+            }
+        }
+    }
+
+    return robj;
+}
+
+void CDataTreeView::AddTextures()
+{
+    for (HashTemplateIterator<StringClass, TextureClass *> textureIter(W3DAssetManager::Get_Instance()->Texture_Hash());

```

Comment:
This would be a bit easier to look at by moving the `HashTemplateIterator<StringClass, TextureClass *> textureIter(W3DAssetManager::Get_Instance()->Texture_Hash()` out of the for loop.

Reply:
fixed.